File: backend/model/kesiswaan.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Admission adalah data pendaftar PPDB (Penerimaan Peserta Didik Baru).
// Status memakai pending, accepted, atau rejected.
type Admission struct {
	ID             uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	Name           string     `gorm:"size:150;not null" json:"name"`
	OriginSchool   string     `gorm:"size:150" json:"origin_school"`
	Gender         string     `gorm:"size:10" json:"gender"`
	Phone          string     `gorm:"size:30" json:"phone"`
	Email          string     `gorm:"size:120" json:"email"`
	Address        string     `gorm:"type:text" json:"address"`
	MajorID        *uuid.UUID `gorm:"type:uuid;index" json:"major_id,omitempty"`
	Status         string     `gorm:"size:20;not null;default:'pending'" json:"status"`
	Note           string     `gorm:"type:text" json:"note"`
	RegisteredAt   *time.Time `json:"registered_at,omitempty"`
	AcademicYearID *uuid.UUID `gorm:"type:uuid;index" json:"academic_year_id,omitempty"`

	Major *Major `gorm:"foreignKey:MajorID" json:"major,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (a *Admission) BeforeCreate(_ *gorm.DB) error {
	if a.ID == uuid.Nil {
		a.ID = uuid.New()
	}
	return nil
}

// StudentCoaching adalah catatan pembinaan kesiswaan pada seorang siswa.
type StudentCoaching struct {
	ID        uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	StudentID uuid.UUID  `gorm:"type:uuid;not null;index" json:"student_id"`
	ClassID   *uuid.UUID `gorm:"type:uuid;index" json:"class_id,omitempty"`
	MajorID   *uuid.UUID `gorm:"type:uuid;index" json:"major_id,omitempty"`
	Topic     string     `gorm:"size:150;not null" json:"topic"`
	Detail    string     `gorm:"type:text" json:"detail"`
	CoachName string     `gorm:"size:120" json:"coach_name"`
	Date      *time.Time `json:"date,omitempty"`

	Student *Student `gorm:"foreignKey:StudentID" json:"student,omitempty"`
	Class   *Class   `gorm:"foreignKey:ClassID" json:"class,omitempty"`
	Major   *Major   `gorm:"foreignKey:MajorID" json:"major,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (s *StudentCoaching) BeforeCreate(_ *gorm.DB) error {
	if s.ID == uuid.Nil {
		s.ID = uuid.New()
	}
	return nil
}

// TalentDevelopment adalah catatan pengembangan bakat dan minat siswa.
type TalentDevelopment struct {
	ID        uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	StudentID uuid.UUID  `gorm:"type:uuid;not null;index" json:"student_id"`
	ClassID   *uuid.UUID `gorm:"type:uuid;index" json:"class_id,omitempty"`
	MajorID   *uuid.UUID `gorm:"type:uuid;index" json:"major_id,omitempty"`
	Field     string     `gorm:"size:100;not null" json:"field"`
	Category  string     `gorm:"size:50" json:"category"`
	Detail    string     `gorm:"type:text" json:"detail"`
	Mentor    string     `gorm:"size:120" json:"mentor"`

	Student *Student `gorm:"foreignKey:StudentID" json:"student,omitempty"`
	Class   *Class   `gorm:"foreignKey:ClassID" json:"class,omitempty"`
	Major   *Major   `gorm:"foreignKey:MajorID" json:"major,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (t *TalentDevelopment) BeforeCreate(_ *gorm.DB) error {
	if t.ID == uuid.Nil {
		t.ID = uuid.New()
	}
	return nil
}

// StudentActivity adalah kegiatan kesiswaan (misalnya ekstrakurikuler atau acara).
type StudentActivity struct {
	ID          uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	Name        string     `gorm:"size:150;not null" json:"name"`
	Type        string     `gorm:"size:50" json:"type"`
	Description string     `gorm:"type:text" json:"description"`
	Organizer   string     `gorm:"size:120" json:"organizer"`
	Location    string     `gorm:"size:150" json:"location"`
	StartDate   *time.Time `json:"start_date,omitempty"`
	EndDate     *time.Time `json:"end_date,omitempty"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (s *StudentActivity) BeforeCreate(_ *gorm.DB) error {
	if s.ID == uuid.Nil {
		s.ID = uuid.New()
	}
	return nil
}

File: frontend/src/lib/kesiswaan.ts
// Tipe modul Kesiswaan.

export interface Admission {
  id: string;
  name: string;
  origin_school: string;
  gender: string;
  phone: string;
  email: string;
  address: string;
  major_id?: string;
  status: string;
  note: string;
  registered_at?: string;
  academic_year_id?: string;
  // student_id terisi setelah pendaftar dikonversi, sekaligus penanda supaya
  // tombol konversi tidak muncul lagi.
  student_id?: string;
  major?: { name: string };
}

export interface StudentCoaching {
  id: string;
  student_id: string;
  class_id?: string;
  major_id?: string;
  topic: string;
  detail: string;
  coach_name: string;
  date?: string;
  student?: { name: string; nis: string };
  class?: { name: string };
  major?: { name: string };
}

export interface TalentDevelopment {
  id: string;
  student_id: string;
  class_id?: string;
  major_id?: string;
  field: string;
  category: string;
  detail: string;
  mentor: string;
  student?: { name: string; nis: string };
  class?: { name: string };
  major?: { name: string };
}

export interface StudentActivity {
  id: string;
  name: string;
  type: string;
  field: string;
  description: string;
  organizer: string;
  location: string;
  start_date?: string;
  end_date?: string;
}

export interface ActivityParticipant {
  id: string;
  activity_id: string;
  student_id: string;
  role: string;
  note: string;
  student?: { name: string; nis: string; class?: { name: string }; major?: { name: string } };
  activity?: { name: string; type: string };
}

export const admissionStatuses = ["pending", "accepted", "rejected"];
export const participantRoles = ["anggota", "ketua", "pembina"];

// bidangOptions dipakai bersama oleh Bakat dan Minat, Kegiatan, dan Prestasi
// supaya ketiganya dapat dicocokkan. Untuk sementara berupa konstanta, dinaikkan
// jadi tabel master bila sekolah perlu mengaturnya sendiri.
export const bidangOptions = [
  "olahraga",
  "seni dan budaya",
  "keagamaan",
  "sains dan teknologi",
  "bahasa dan sastra",
  "kepemimpinan dan organisasi",
  "kewirausahaan",
  "sosial dan kemanusiaan",
  "lainnya",
];
